File: go/src/core/agent.go
package core

import (
	"fmt"

	"minicc/src/core/providers"
	"minicc/src/tools"
)

type Agent struct {
	provider providers.LLMProvider
}

func NewAgent(provider providers.LLMProvider) *Agent {
	return &Agent{provider: provider}
}

func (a *Agent) handleToolCalls(toolCalls []map[string]interface{}) []map[string]interface{} {
	var results []map[string]interface{}

	for _, call := range toolCalls {
		id := call["id"].(string)
		name := call["name"].(string)
		args := call["args"].(map[string]interface{})

		if args["_parse_error"] != nil {
			results = append(results, map[string]interface{}{
				"id":      id,
				"name":    name,
				"result":  fmt.Sprintf("[Agent 内部错误] 你输出的工具参数 JSON 格式不合法，无法解析。\n你输出的原始参数为:\n%s", args["_raw_arguments"]),
				"isError": true,
			})
			continue
		}

		var tool *tools.Tool
		for _, t := range tools.Tools {
			if t.Name == name {
				tool = &t
				break
			}
		}

		if tool == nil {
			fmt.Printf("\033[31m[Agent] 未知工具: %s\033[0m\n", name)
			results = append(results, map[string]interface{}{
				"id":      id,
				"name":    name,
				"result":  fmt.Sprintf("未知的工具调用: %s", name),
				"isError": true,
			})
			continue
		}

		fmt.Printf("\033[36m▶ [Agent] 正在调用工具: %s ...\033[0m\n", name)
		result, err := tool.Execute(args)
		fmt.Printf("\033[32m✔ [Agent] 工具 %s 执行完毕。\033[0m\n", name)

		if len(result) > 8000 {
			fmt.Printf("\n[上下文瘦身] 工具 %s 返回结果过长 (%d 字符)，已触发截断。\n", name, len(result))
			result = result[:8000] + "\n\n...[由于内容过长，已被截断]..."
		}

		if err != nil {
			results = append(results, map[string]interface{}{
				"id":      id,
				"name":    name,
				"result":  fmt.Sprintf("执行工具 %s 时出错: %v", name, err),
				"isError": true,
			})
		} else {
			results = append(results, map[string]interface{}{
				"id":      id,
				"name":    name,
				"result":  result,
				"isError": false,
			})
		}
	}

	return results
}

func (a *Agent) Chat(userMessage string, onTextResponse func(text string, isThinking bool)) {
	response, err := a.provider.SendMessage(userMessage, onTextResponse)
	if err != nil {
		fmt.Printf("\n[Agent 报错] %v\n", err)
		return
	}

	loopCount := 0
	maxLoops := 5

	for {
		toolCallsRaw := response["toolCalls"]
		if toolCallsRaw == nil {
			break
		}

		toolCalls := toolCallsRaw.([]map[string]interface{})
		if len(toolCalls) == 0 {
			break
		}

		loopCount++
		if loopCount > maxLoops {
			fmt.Printf("\n\033[33m[Agent] 工具调用循环次数过多 (%d)，为防止无限循环，已强制终止。\033[0m\n", loopCount)
			break
		}

		fmt.Printf("\n\033[33m[Agent] 收到大模型指令，准备执行 %d 个工具调用... (第 %d 轮)\033[0m\n", len(toolCalls), loopCount)
		toolResults := a.handleToolCalls(toolCalls)

		fmt.Printf("\n\033[33m[Agent] 工具执行完毕，正在将结果发送回大模型，请稍候...\033[0m\n\n")
		response, err = a.provider.SendToolResults(toolResults, onTextResponse)
		if err != nil {
			fmt.Printf("\n[Agent 报错] %v\n", err)
			return
		}
	}
}

File: rust/src/core/agent.rs
use crate::core::providers::{LLMProvider, OnTextResponseFn, ToolCall, ToolResult};
use crate::tools::get_all_tools;

pub struct Agent {
    provider: Box<dyn LLMProvider>,
}

impl Agent {
    pub fn new(provider: Box<dyn LLMProvider>) -> Self {
        Agent { provider }
    }

    async fn handle_tool_calls(&self, tool_calls: Vec<ToolCall>) -> Vec<ToolResult> {
        let mut results = Vec::new();

        for call in tool_calls {
            let id = call.id;
            let name = call.name;
            let args = call.args;

            if let Some(err) = args.get("_parse_error") {
                if err.as_bool().unwrap_or(false) {
                    let raw = args.get("_raw_arguments").and_then(|v| v.as_str()).unwrap_or("");
                    results.push(ToolResult {
                        id,
                        name: name.clone(),
                        result: format!("[Agent 内部错误] 你输出的工具参数 JSON 格式不合法，无法解析。\n你输出的原始参数为:\n{}", raw),
                        is_error: true,
                    });
                    continue;
                }
            }

            let tools = get_all_tools();
            let tool = tools.into_iter().find(|t| t.name == name);

            match tool {
                Some(t) => {
                    println!("\x1b[36m▶ [Agent] 正在调用工具: {} ...\x1b[0m", name);
                    match (t.execute)(args).await {
                        Ok(mut result) => {
                            println!("\x1b[32m✔ [Agent] 工具 {} 执行完毕。\x1b[0m", name);
                            if result.len() > 8000 {
                                println!("\n[上下文瘦身] 工具 {} 返回结果过长 ({} 字符)，已触发截断。", name, result.len());
                                result = format!("{}\n\n...[由于内容过长，已被截断]...", &result[0..8000]);
                            }
                            results.push(ToolResult {
                                id,
                                name,
                                result,
                                is_error: false,
                            });
                        }
                        Err(e) => {
                            results.push(ToolResult {
                                id,
                                name: name.clone(),
                                result: format!("执行工具 {} 时出错: {}", name, e),
                                is_error: true,
                            });
                        }
                    }
                }
                None => {
                    println!("\x1b[31m[Agent] 未知工具: {}\x1b[0m", name);
                    results.push(ToolResult {
                        id,
                        name: name.clone(),
                        result: format!("未知的工具调用: {}", name),
                        is_error: true,
                    });
                }
            }
        }

        results
    }

    pub async fn chat(&mut self, user_message: String, on_text_response: OnTextResponseFn) {
        match self.provider.send_message(user_message, &on_text_response).await {
            Ok(mut response) => {
                let mut loop_count = 0;
                let max_loops = 5;

                while !response.tool_calls.is_empty() {
                    loop_count += 1;
                    if loop_count > max_loops {
                        println!("\n\x1b[33m[Agent] 工具调用循环次数过多 ({})，为防止无限循环，已强制终止。\x1b[0m", loop_count);
                        break;
                    }

                    println!("\n\x1b[33m[Agent] 收到大模型指令，准备执行 {} 个工具调用... (第 {} 轮)\x1b[0m", response.tool_calls.len(), loop_count);
                    let tool_results = self.handle_tool_calls(response.tool_calls).await;

                    println!("\n\x1b[33m[Agent] 工具执行完毕，正在将结果发送回大模型，请稍候...\x1b[0m\n");
                    match self.provider.send_tool_results(tool_results, &on_text_response).await {
                        Ok(res) => response = res,
                        Err(e) => {
                            println!("\n[Agent 报错] {}", e);
                            break;
                        }
                    }
                }
            }
            Err(e) => println!("\n[Agent 报错] {}", e),
        }
    }
}